### PR TITLE
[WIP][FSDP2] post_backward idempotency guard — does not fix the underlying AOT-aliasing bug

### PR DESCRIPTION
**Pull request by TimDarcet**
> **UPDATE 2026-07-24: this patch has no effect on the bug.** Verified in source: `post_backward` fires **exactly once** in the buggy run. There is no double-reduce. The idempotency guard here is defensively fine but does not fix the observed grad-norm inflation. The real mechanism is a dynamo/AOT-autograd tensor-aliasing miswiring — the compiled backward feeds the wrong slice tensor into `F.cross_entropy`'s `target` argument when two forward inputs share storage. Keeping this PR as a draft while the root cause is investigated upstream. See comment below for the walkthrough. Original body preserved below.

## Status: WIP — needs rework

Companion to #190938 (also being reworked). Neither patch as-currently-authored addresses the actual bug.

## What we now believe the bug is

`torch.compile` + input tensors that share storage produce a silent-correctness backward under nightly `2.14.0.dev20260722+cu130`. FSDP2 / CP / `ModuleList` are not required — they were just ingredients that put dynamo on the codepath where the miswiring occurs.

Repro (23 lines, single GPU, no FSDP, no CP):

```python
import torch, torch.nn as nn, torch.nn.functional as F
D, S = 512, 512

def _sl(x): return x

class B(nn.Module):
    def __init__(s):
        super().__init__(); s.i = nn.Linear(D, D, bias=False)
    def forward(s, x): return s.i(x)

class M(nn.Module):
    def __init__(s):
        super().__init__()
        s.e = nn.Embedding(D, D)
        s.b = nn.ModuleList([B()])
    def forward(s, tok, tgt):
        h = s.e(_sl(tok.flatten())).view(1, -1, D)
        for blk in s.b:
            h = blk(h)
        return F.cross_entropy(h.reshape(-1, D), _sl(tgt.flatten()), reduction="none")

torch.manual_seed(0)
m = M().cuda()
for p in m.parameters(): p.data.normal_(0, 0.02)
m.compile()
torch.manual_seed(1)
r = ((torch.rand(1, S+1) ** 8) * D).long().cuda()
loss = m(r[:, :-1], r[:, 1:]).float().mean()   # tok and tgt share storage with r
loss.backward()
print(f"loss={loss.item():.6f} grad e.weight:{m.e.weight.grad.float().norm():.4f}")
```

Cloning the two slices before the call is bit-perfect:

```python
tok = r[:, :-1].clone()
tgt = r[:, 1:].clone()
loss = m(tok, tgt).float().mean()
```

| Variant                  | loss     | grad `e.weight` |
|--------------------------|----------|-----------------|
| aliased inputs, compile  | 6.242074 | 0.1992 (BUG)    |
| aliased inputs, eager    | 6.239221 | 0.0863          |
| cloned inputs, compile   | 6.239221 | 0.0863 (FIXED)  |
| cloned inputs, eager     | 6.239221 | 0.0863          |

## Why this PR's patch has no effect

- Instrumentation of `FSDPParamGroup.post_backward` shows it fires **exactly once** in the buggy run.
- `RegisterPostBackwardFunction.apply` / `.backward` and `AccumulateGrad` each fire once.
- The compiled backward FX graph has a single `mm` per weight, into a fresh output buffer. Nothing is being reduced twice.
- The `tangents_1.norm()` handed to the compiled backward matches eager (0.062440 vs 0.062439). But **element-wise** the tangent is a row-shifted version of the eager tangent: recovering the target-index sequence from it produces the `tok` slice, not the `tgt` slice. So the compiled forward is feeding the wrong slice-graph output into `F.cross_entropy`'s `target`.

An idempotency guard on `post_backward` is a fine sanity-check to keep around, but it is not what fixes this bug.

## Where a real fix likely goes

`torch/_functorch/_aot_autograd/*` — input-alias / view-tracking. Possibly with a contributing bug in `torch/_dynamo/variables/tensor.py` in how views over a shared base storage are exposed to AOT. Investigation ongoing.

## Temporary user-side mitigation

`.clone()` input tensor pairs that share storage before calling a compiled module.

## References

- Umbrella issue: #142358
- Companion PR (also being reworked): #190938 (dynamo `gb7000` recursive-SKIP — was the previous best guess at the root cause; now known to only workaround the AOT miswiring by forcing eager).

---

<details>
<summary>Original PR body (misdiagnosis, kept for context)</summary>

**LLM Statement:** This is in most part the work of my Claude, both the bug diagnostic, the narrowing down to a minimal reprod, the suggestion of a fix, and the body of this PR where done by it.

## Summary

**Companion patch to the dynamo loop-break fix** (which addresses the root cause: recursive SKIP propagation for `gb7000` graph-break-in-loop). This PR adds defense-in-depth on the FSDP2 side by guarding `post_backward` against being invoked twice within a single iteration.

(rest of original body elided — see PR history.)

</details>

### CODE REVIEW

**Comment by TimDarcet** [comment]
Update — this PR's patch does not fix the bug. Marking as WIP and reworking the body.

## What I verified in-source

I applied PATCH 2 (this PR's `_already_reduced_this_iter` guard) to `torch/distributed/fsdp/_fully_shard/_fsdp_param_group.py` in-place on nightly `2.14.0.dev20260722+cu130`, cleared `__pycache__`, and reran the repro. **Grad norms remained inflated identically** to unpatched. Reverted the source, re-checked — same buggy numbers under `torch.compile`. The guard is behaviorally inert on this bug.

## What's actually happening (evidence)

- `FSDPParamGroup.post_backward` fires **exactly once** per group per iter in the buggy run. The "double-reduce" premise this PR was built on is wrong.
- `RegisterPostBackwardFunction.apply` and `.backward` fire once. `AccumulateGrad` fires once per param.
- The compiled backward FX graph is clean: single `mm` per weight, fresh output buffer.
- `tangents_1.norm()` matches eager (0.062440 vs 0.062439) — the loss's own gradient magnitude is right.
- **But element-wise**, the compiled tangent is a **row-shifted** version of the eager tangent. Decoding the target-index sequence from the tangent shows that under compile, `F.cross_entropy` receives the `tok` slice where it should be getting the `tgt` slice.
- The bug reproduces with **no FSDP2 and no CP** — see #190938 for the reduced 23-line repro.
- Cloning the two aliased slice inputs (`tok = r[:, :-1].clone(); tgt = r[:, 1:].clone()`) fixes it bit-for-bit under compile.

This is a dynamo/AOT-autograd tensor-aliasing miswiring, orthogonal to FSDP2. FSDP2's role in the original repro was just to compose the specific compilation path where AOT mishandles the aliased slice-graph outputs.

## Plan

- Keeping this PR as a draft — the guard is a reasonable defense-in-depth even though it doesn't fix this bug, and I want the investigation trail preserved.
- The real fix will most likely land in `torch/_functorch/_aot_autograd/*` (input-alias / dedup / view-tracking), possibly with a contributing patch in `torch/_dynamo/variables/tensor.py`. I don't have a proposed fix yet.
- Umbrella: #142358. Companion PR: #190938.

## Temporary user-side mitigation

`.clone()` input tensor pairs that share storage before calling a compiled module.

**Comment by TimDarcet** [comment]
The **actual root-cause fix** is at #191002. Nothing double-fires in FSDP2 — the bug is inside AOT-autograd's `gen_alias_from_base` when inductor clones a misaligned input onto fresh storage. Verified. This PR's idempotency guard has no effect and can be closed once #191002 lands.
